ipfsapi/multipart.py

```diff
@@ -328,20 +377,30 @@ class DirectoryStream(BufferedGenerator):
     ----------
     directory : str
         The filepath of the directory to encode
+    patterns : str | list
+        A single glob pattern or a list of several glob patterns and
+        compiled regular expressions used to determine which filepaths to match
     chunk_size : int
         The maximum size that any single file chunk may have in bytes
     """
 
     def __init__(self,
                  directory,
                  recursive=False,
-                 fnpattern='*',
+                 patterns='**',
                  chunk_size=default_chunk_size):
         BufferedGenerator.__init__(self, directory, chunk_size=chunk_size)
 
-        self.directory = directory
+        self.patterns = []
+        patterns = [patterns] if isinstance(patterns, str) else patterns
+        for pattern in patterns:
+            if isinstance(pattern, str):
+                self.patterns.append(glob_compile(pattern))
+            else:
```

Comment:
Allowing *anything that implements the `.match` "interface"* here is indeed considered "Pythonic", it also allows programmers to implement custom matching function that do not use `re` at all. (Note however that the term Pythonic isn't very well defined and highly subjective; other people may interpret it very differently.)
Also, from a practical point-of-view, it very hard to reliably detect whether something actually is a compiled regular expression object or not, since `type(re.compile(""))` returns `'_sre.SRE_Pattern'`, which is (1) private (starts with an underscore) and (2) virtual (cannot be accessed directly from Python code: There is no `_sre.SRE_Pattern` class accessible from Python).